src/webui/api/searchcontroller.cpp

```diff
@@ -0,0 +1,347 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "searchcontroller.h"
+
+#include <QRegularExpression>
+#include <QSharedPointer>
+#include <QTimer>
+
+#include "base/logger.h"
+#include "base/search/searchhandler.h"
+#include "base/utils/misc.h"
+#include "base/utils/random.h"
+#include "base/utils/string.h"
+#include "apierror.h"
+#include "isessionmanager.h"
+
+class SearchPluginManager;
+
+using SearchHandlerPtr = QSharedPointer<SearchHandler>;
+using SearchHandlerDict = QHash<QString, SearchHandlerPtr>;
+
+void SearchController::startAction()
+{
+    checkParams({"pattern", "category", "plugins"});
+
+    if (Utils::Misc::pythonVersion() < 0)
+        throw APIError(APIErrorType::Conflict, "Python must be installed to use the Search Engine.");
+
+    const QString pattern = params()["pattern"].trimmed();
+    const QString category = params()["category"].trimmed();
+    const QString plugins = params()["plugins"].trimmed();
+    const QString pluginsLower = plugins.toLower();
+
+    QStringList pluginsToUse;
+    if (pluginsLower == "all")
+        pluginsToUse = SearchPluginManager::instance()->allPlugins();
+    else if ((pluginsLower == "enabled") || (pluginsLower == "multi"))
+        pluginsToUse = SearchPluginManager::instance()->enabledPlugins();
+    else
+        pluginsToUse << plugins;
+
+    const QString id = generateSearchId();
+
+    auto searchHandlers = sessionManager()->session()->getData(QLatin1String("searchHandlers")).value<SearchHandlerDict >();
+    SearchHandlerPtr searchHandler {SearchPluginManager::instance()->startSearch(pattern, category, pluginsToUse)};
+    searchHandlers.insert(id, searchHandler);
+    sessionManager()->session()->setData(QLatin1String("searchHandlers"), QVariant::fromValue(searchHandlers));
+
+    const int thirtyMinutes = 1800000;
+    QTimer::singleShot(thirtyMinutes, this, [id, this] () {searchTimeout(id);});
```

Comment:
This is to clean up the search job after a timeout. We don't want all jobs around for the life of the session, do we? If this isn't a concern I can remove it.

Reply:
> It can be a problem in case of wrong use.

We can't rely on the web session calling delete on each search job. I don't foresee the webui doing this at all unless the user takes explicit action (e.g. closing the search tab), which very likely wouldn't happen.

I think a clean up timer is our best option. I don't think a max jobs limit solves for this same issue. Max jobs would prevent someone from maliciously spawning a bunch of jobs to try using all system resources. If a user has a long-lived webui session, and does many searches over time, should they suddenly no longer be allowed to run a new search?

Reply:
> Why? What do you think it's for?

It exists in the API for obvious reasons, but doesn't cleanly fit into the actual webui usage. Unless the user takes an explicit action, like closing a search tab (as the current GUI `SearchWidget` does). In general I don't expect a user to close their search tabs- they may just close the browser tab entirely.

> When the session ends it destroys all session data.

Of course, this is still a valid way to delete the search jobs. But we should also account for long-lived sessions.

> Did I suggest to prevent it from start a new search? Apparently, it wasn't clear enough... I meant delete the old search when the user starts a new one and the limit is exceeded.

Thanks for clarifying. I still fail to see why we would set an arbitrary limit on this, when this limit doesn't also exist in the GUI (please correct me if this is wrong). My goal with the timer is to *eventually* delete the search job when the user session is long-lived and the user doesn't explicitly delete it. Maybe all this is unnecessary and we just wait until the session is deleted to delete all search jobs.

Reply:
I've increased this timeout to 1 hour. I still firmly believe a timer is a good idea, but also welcome other input. @Chocobo1 @sledgehammer999 

Web sessions are never deleted until the user clicks log out. Which I strongly doubt any users actually do.

Reply:
>Web sessions are never deleted until the user clicks log out. Which I strongly doubt any users actually do.

@Piccirello, if you don't know something and you're not able to learn it from the code yourself, just ask next time. Session inactive time is 15 minutes. It will be deleted after (in fact, it won't be deleted exactly in 15 minutes, as I avoid (over-)using timers where possible; actually it will be deleted on next sessions-related action, i.e. when user access this (outdated) session or try to start the new one).
Sessions ensure that resources are freed when the session is closed (by the user or as a result of a timeout). I agree that this is not enough in some cases. Each search start causes the allocation of resources, so we need to make sure that it is not excessive.
But what do we really want to prevent? What's wrong with having a user have one or more active searches throughout the session (even if it's active for a few hours)? On the other hand, what can we prevent by having a timeout for each search? Nothing! How many searches can I run during this time? Not even during hour, but just during couple of minutes?..
So I repeat again, we should limit the number of existing searches instead.

Reply:
>  Session inactive time is 15 minutes. It will be deleted after

Could you point me to where this happens in the code? I couldn't find it when I looked last night.

> What's wrong with having a user have one or more active searches throughout the session (even if it's active for a few hours)?

If the session is cleared so aggressively then I'm less concerned about it. I should disable my IP whitelist rule so that I can see how often the session is recreated. It just happens in the background for me, without any auth prompt, so it appears to be one long session.

> What's wrong with having a user have one or more active searches throughout the session (even if it's active for a few hours)?

The concern was around a long-lived (e.g. multi-day) session. I'll test to confirm, but this seems more unlikely now.

> So I repeat again, we should limit the number of existing searches instead.

I still don't think this is the right approach. The concern isn't concurrent sessions, it's sessions not being cleared over a long time period.

Reply:
>I still don't think this is the right approach. 

I.e. do you want to allow a remote user to start unlimited number of searches?
>The concern isn't concurrent sessions, it's sessions not being cleared over a long time period.

Why should session be cleared when it's active (i.e. actually used)?

Please provide me an example of use case when long lived "searches" bring real problems (assuming there is max searches count limit exists).

Reply:
>Could you point me to where this happens in the code? I couldn't find it when I looked last night.

You can search for INACTIVE_TIME constant usages in WebApplication code.

Reply:
> I.e. do you want to allow a remote user to start unlimited number of searches?

What limit would you impose?

> Why should session be cleared when it's active (i.e. actually used)?

My point is that a search job may exist long after it's actually "active," simply because the webui does not provide a good means of setting the search job as inactive. This is just a byproduct of how people use webpages.

I really don't care to belabor this point any longer. If someone else wants to provide input, please do. Otherwise I plan to have a search job timer (following @Chocobo1's suggestion to use a single timer).